File: lib/check/scoring.js
const CATEGORY_MAP = {
  scope: "scope",
  validation: "validation",
  consistency: "consistency",
  continuity: "continuity",
  performance: "performance",
  risk: "risk"
};

const SEVERITY_PENALTY = {
  error: 15,
  warning: 5
};

export function computeCompositeScore(findings, weights) {
  if (!findings || findings.length === 0) {
    return 100;
  }

  const effectiveWeights = (weights && typeof weights === "object" && Object.keys(weights).length > 0)
    ? weights
    : null;

  let totalWeightedDeduction = 0;
  let totalWeight = 0;

  for (const finding of findings) {
    const basePenalty = SEVERITY_PENALTY[finding.severity] ?? 0;
    const category = CATEGORY_MAP[finding.category] || finding.category;

    if (effectiveWeights && effectiveWeights[category] != null) {
      const w = effectiveWeights[category] / 100;
      totalWeightedDeduction += basePenalty * w;
      totalWeight += w;
    } else {
      totalWeightedDeduction += basePenalty;
      totalWeight += 1;
    }
  }

  if (totalWeight > 0) {
    totalWeightedDeduction = totalWeightedDeduction * (findings.length / totalWeight);
  }

  return Math.max(0, Math.min(100, Math.round((100 - totalWeightedDeduction) * 10) / 10));
}

export function getScoreVerdict(score, hasErrors) {
  if (hasErrors) {
    return "blocked";
  }
  if (score >= 90) {
    return "safe-to-deploy";
  }
  if (score >= 70) {
    return "pass-with-concerns";
  }
  if (score >= 40) {
    return "needs-attention";
  }
  return "high-risk";
}

export function validateScoringWeights(weights) {
  const sum = Object.values(weights).reduce((total, w) => total + w, 0);
  return sum === 100;
}

export function formatScoreBar(score, verdict, locale = "en") {
  const filled = Math.round(score / 10);
  const empty = 10 - filled;
  const bar = "█".repeat(filled) + "░".repeat(empty);
  const label = locale === "zh-CN" ? "信任评分" : "Trust Score";
  return `${label}: ${bar} ${score}/100 (${verdict})`;
}

File: crates/agent-guardrails-cli/src/check/scoring.rs
use super::Finding;
use std::collections::HashMap;

pub fn compute_composite_score(
    findings: &[Finding],
    weights: Option<&HashMap<String, f64>>,
) -> f64 {
    if findings.is_empty() {
        return 100.0;
    }

    let effective_weights = weights.filter(|value| !value.is_empty());
    let mut total_weighted_deduction = 0.0;
    let mut total_weight = 0.0;

    for finding in findings {
        let base_penalty = severity_penalty(&finding.severity);
        let category = mapped_category(&finding.category);

        if let Some(weights) = effective_weights {
            if let Some(weight) = weights.get(category) {
                let normalized_weight = weight / 100.0;
                total_weighted_deduction += base_penalty * normalized_weight;
                total_weight += normalized_weight;
                continue;
            }
        }

        total_weighted_deduction += base_penalty;
        total_weight += 1.0;
    }

    if total_weight > 0.0 {
        total_weighted_deduction *= findings.len() as f64 / total_weight;
    }

    round_one_decimal((100.0 - total_weighted_deduction).clamp(0.0, 100.0))
}

pub fn get_score_verdict(score: f64, has_errors: bool) -> &'static str {
    if has_errors {
        return "blocked";
    }
    if score >= 90.0 {
        return "safe-to-deploy";
    }
    if score >= 70.0 {
        return "pass-with-concerns";
    }
    if score >= 40.0 {
        return "needs-attention";
    }
    "high-risk"
}

fn severity_penalty(severity: &str) -> f64 {
    match severity {
        "error" => 15.0,
        "warning" => 5.0,
        _ => 0.0,
    }
}

fn mapped_category(category: &str) -> &str {
    match category {
        "scope" => "scope",
        "validation" => "validation",
        "consistency" => "consistency",
        "continuity" => "continuity",
        "performance" => "performance",
        "risk" => "risk",
        other => other,
    }
}

fn round_one_decimal(value: f64) -> f64 {
    (value * 10.0).round() / 10.0
}
